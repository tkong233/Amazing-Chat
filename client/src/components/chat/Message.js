import React from "react";
import { connect } from "react-redux";
import Popover from "@material-ui/core/Popover";
import Typography from "@material-ui/core/Typography";
import { makeStyles } from "@material-ui/core/styles";

import { deleteMessage } from "../../actions/chatActions";
import "./Message.css";
import { MessageBox } from "react-chat-elements";
import "react-chat-elements/dist/main.css";

const useStyles = makeStyles((theme) => ({
  typography: {
    padding: theme.spacing(2),
  },
}));

const Message = (props) => {
  // const { text, sender, user, type, date } = props;
  const {
    text,
    sender,
    user,
    type,
    from,
    to,
    pairId,
    datetime,
    message,
  } = props;
  const dateTimeZone = new Date(datetime).toString();
  const classes = useStyles();
  const position = sender === user ? "right" : "left";
  const anchorOrigin = {
    vertical: "top",
    horizontal: position,
  };
  const style =
    position === "left" ? { textAlign: "left" } : { textAlign: "right" };

  const [anchorEl, setAnchorEl] = React.useState(null);

  const handleOpenPopover = (event) => {
    setAnchorEl(event.currentTarget);
  };

  const handleClosePopover = () => {
    setAnchorEl(null);
  };

  const open = Boolean(anchorEl);
  const id = open ? "simple-popover" : undefined;

  const handleMessageOnClick = (e) => {
    handleOpenPopover(e);
  };

  const handleDeleteMessage = () => {
    console.log("handle delete message", message);
    const data = { from, to, message, pairId, datetime };
    props.deleteMessage(data);
    handleClosePopover();
  };

  if (
    props.text.substring(0, 45) ===
      "https://cis557-amazing-chat.s3.amazonaws.com/" ||
    "uploads/"
  ) {
    const file_ext = props.text
      .substring(props.text.length - 3, props.text.length)
      .toLowerCase();
    if (
      file_ext === "mp4" ||
      file_ext === "webm" ||
      file_ext === "avi" ||
      file_ext === "mov"
    ) {
      return (
        <div style={style}>
          <video
            style={{
              marginLeft: "20px",
              marginRight: "20px",
              marginTop: "5px",
              marginBottom: "5px",
              borderRadius: "8px",
              boxShadow: "0 0 2px 0px gray",
              maxWidth: "300px",
            }}
            src={`${props.text}`}
            alt="video"
            type="video/mp4"

            controls
            onClick={handleMessageOnClick}
          />
          <Popover
            id={id}
            open={open}
            anchorEl={anchorEl}
            onClose={handleClosePopover}
            anchorOrigin={anchorOrigin}
            transformOrigin={{
              vertical: "top",
              horizontal: "center",
            }}
          >
            <Typography
              className={classes.typography}
              onClick={handleDeleteMessage}
            >
              Delete
            </Typography>
          </Popover>
        </div>
      );
    } else if (file_ext === "mp3" || file_ext === "wav" || file_ext === "ogg") {
      return (
        <div style={style}>
          <audio
            src={`${props.text}`}
            alt="audio"
            type="audio/ogg"
            style={{
              marginLeft: "20px",
              marginRight: "20px",
              marginTop: "5px",
              marginBottom: "5px",
              borderRadius: "8px",
              boxShadow: "0 0 2px 0px gray",
            }}
            controls
            onClick={handleMessageOnClick}
          />
          <Popover
            id={id}
            open={open}
            anchorEl={anchorEl}
            onClose={handleClosePopover}
            anchorOrigin={anchorOrigin}
            transformOrigin={{
              vertical: "top",
              horizontal: "center",
            }}
          >
            <Typography
              className={classes.typography}
              onClick={handleDeleteMessage}
            >
              Delete
            </Typography>
          </Popover>
        </div>
      );
    } else if (
      file_ext === "png" ||
      file_ext === "jpg" ||
      file_ext === "jpeg"
    ) {
      return (
        <div style={style}>
          <img
            // max-width = "250px"
            // max-height = "250px"
            src={`${props.text}`}
            alt="Imgupload"
            style={{
              marginLeft: "20px",
              marginRight: "20px",
              marginTop: "5px",
              marginBottom: "5px",
              maxWidth: "250px",
              maxHeight: "250px",
              borderRadius: "8px",
              boxShadow: "0 0 2px 0px gray",
            }}
            onClick={handleMessageOnClick}
          />
          <Popover
            id={id}
            open={open}
            anchorEl={anchorEl}
            onClose={handleClosePopover}
            anchorOrigin={anchorOrigin}
            transformOrigin={{
              vertical: "top",
              horizontal: "center",
            }}
          >
            <Typography
              className={classes.typography}
              onClick={handleDeleteMessage}
            >
              Delete
            </Typography>
          </Popover>
        </div>
        // <div>
        //   <MessageBox
        //   className="message-bubble"
        //   position={position}
        //   type={"photo"}
        //   data={{
        //     uri: `${props.text}`,
        //   }}
        // />
        // </div>
      );
    } else {
      return (
        <div>
          <MessageBox
            className="message-bubble"
            position={position}
            type={"text"}
            text={props.text}
            dateString={
              dateTimeZone.substring(3, 11) + dateTimeZone.substring(15, 24)
            }
            onClick={handleMessageOnClick}
          />
          <Popover
            id={id}
            open={open}
            anchorEl={anchorEl}
            onClose={handleClosePopover}
            anchorOrigin={anchorOrigin}
            transformOrigin={{
              vertical: "top",
              horizontal: "center",
            }}
          >
            <Typography
              className={classes.typography}
              onClick={handleDeleteMessage}
            >
              Delete
            </Typography>
          </Popover>
        </div>
      );
    }
  } else {
    return (
      <div data-test="MessageComponent">
        <MessageBox
          className="message-bubble"
          position={position}
          type={type}
          text={text}
          onClick={handleMessageOnClick}
        />
        <Popover
          id={id}
          open={open}
          anchorEl={anchorEl}
          onClose={handleClosePopover}
          anchorOrigin={anchorOrigin}
          transformOrigin={{
            vertical: "top",
            horizontal: "center",
          }}
        >
          <Typography
            className={classes.typography}
            onClick={handleDeleteMessage}
          >
            Delete
          </Typography>
        </Popover>
      </div>
    );
  }
};

const mapStateToProps = (state) => ({
  
});

export default connect(mapStateToProps, { deleteMessage })(Message);
