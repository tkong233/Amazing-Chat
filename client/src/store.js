import { createStore, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import { persistStore, persistReducer } from 'redux-persist'
import rootReducer from "./reducers";
import storage from 'redux-persist/lib/storage' // defaults to localStorage for web

const initialState = {};

const persistConfig = {
  key: 'root',
  storage,
}

const middleware = [thunk];

const persistedReducer = persistReducer(persistConfig, rootReducer, initialState);

const store = createStore(

  persistedReducer,
  compose(
    applyMiddleware(...middleware),
    (window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ &&
      window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__()) ||
      compose
  )
);

const persistor = persistStore(store);

export {
  store,
  persistor,
  middleware
};



