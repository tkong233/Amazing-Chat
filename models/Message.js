const mongoose = require("mongoose");
const Schema = mongoose.Schema;

const MessageSchema = new Schema({
  pairId: {
    type: mongoose.Schema.Types.ObjectId,
    required: true,
  },
  from: {
    type: String,
    required: true
  },
  to: {
    type: String,
    required: true
  },
  message: {
    type: String,
    required: true
  },
  datetime: {
    type: Date,
    default: Date.now
  },
  type:{
    type: String
  }
});

module.exports = Message = mongoose.model("messages", MessageSchema);