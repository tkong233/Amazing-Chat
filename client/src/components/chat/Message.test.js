import React from 'react';
import Message from './Message';
import {shallow} from 'enzyme';
import checkPropTypes from 'check-prop-types';
import Enzyme from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-16'; 
import { applyMiddleware, createStore } from 'redux';
import {middleware} from '../../store';
import rootReducer from '../../reducers/index'

Enzyme.configure({
    adapter: new EnzymeAdapter(),
});

const setUp = (initialState={}) =>{
    const props = {
        text: 'hello hello'
    }
    const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
    const store = createStoreWithMiddleware(rootReducer, initialState);
    const wrapper = shallow(<Message store={store} {...props}/>).dive();
    // console.log(wrapper.debug());
    return wrapper;
}

describe('Message Component', ()=>{
    describe('Checking PropTypes', ()=>{
        it('Should not throw a warning', ()=>{
            const expectedProps = {
                text: '',
                sender: '',
                user: '',
                type: '',
                from: '',
                to: '',
                pairId: '',
                datetime: '',
                message: []
            }
            // eslint-disable-next-line react/forbid-foreign-prop-types
            const propsErr = checkPropTypes(Message.propTypes, expectedProps, 'props', Message.name);
            expect(propsErr).toBeUndefined();
        })
    });
    describe('Renders', ()=>{
        let wrapper;
        beforeEach(()=>{
            const initialState = {
                chat:{
                    socket: 'testsocket'
                }
            };
            wrapper = setUp(initialState);
            wrapper.setProps({text: 'hello hello'})
        });
        
        it('Should render without errors', ()=>{
            const container = wrapper.find(`[data-test='MessageComponent']`);
            expect(container.length).toBe(0);
        });

        it('Snapshot testing', ()=>{
            expect(wrapper).toMatchSnapshot();
        });
        

    });
});
