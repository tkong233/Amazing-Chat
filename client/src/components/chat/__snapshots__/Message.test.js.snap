// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message Component Renders Snapshot testing 1`] = `ShallowWrapper {}`;
